File: android/src/main/java/com/dodo/flutterbridge/call/strategy/SingleHandlerStrategy.kt
package com.dodo.flutterbridge.call.strategy

import com.dodo.flutterbridge.call.Handler
import com.dodo.flutterbridge.call.exception.HandlerNotFoundException
import com.dodo.flutterbridge.call.strategy.SingleHandlerStrategy.ConflictType

/**
 *     author : liuduo
 *     e-mail : liuduo@gyenno.com
 *     time   : 2022/08/11
 *     desc   : 同一个名字只保存单个handler,可设置冲突处理方式
 *     version: 1.0
 */
class SingleHandlerStrategy<A>(
    /**
     * 如果名字重复有冲突处理方式[ConflictType.Replace]替换,和[ConflictType.Ignore]忽略,和[ConflictType.Exception]异常
     */
    private val conflictType: ConflictType
) : StickyHandlerStrategy<A>() {
    private val handlers = mutableMapOf<String, Handler<A>>()

    override fun addHandler(handler: Handler<A>) {
        super.addHandler(handler)
        if (handlers[handler.name] == null) {
            handlers[handler.name] = handler
        } else {
            conflictType(handlers, handler)
        }
    }

    override fun removeHandler(handler: Handler<A>) {
        handlers.remove(handler.name)
    }

    @Throws(HandlerNotFoundException::class)
    override fun onCallStrategy(name: String, data: A): Any? {
        return handlers[name]?.run {
            return onCall(data)
        } ?: throw  HandlerNotFoundException()
    }

    enum class ConflictType {
        /**
         * 替换
         */
        Replace,

        /**
         * 忽略
         */
        Ignore,

        /**
         * 异常
         */
        Exception;

        operator fun <A> invoke(
            handlers: MutableMap<String, Handler<A>>,
            handler: Handler<A>
        ) {
            when (this) {
                Replace -> handlers[handler.name] = handler
                Ignore -> {}
                Exception -> throw Exception("同名字的handler已存在")
            }
        }
    }
}

File: lib/src/call/strategy/single_handler_strategy.dart
import 'package:flutter_bridge/src/call/exception/handler_not_found_exception.dart';
import 'package:flutter_bridge/src/call/handler.dart';
import 'package:flutter_bridge/src/call/strategy/sticky_handler_strategy.dart';

///   author : liuduo
///   e-mail : liuduo@gyenno.com
///   time   : 2022/08/27
///   desc   : 同一个名字只保存单个handler,可设置冲突处理方式
///   version: 1.0
class SingleHandlerStrategy<A> extends StickyHandlerStrategy<A> {
  final Map<String, Handler<A>> _handlers = {};
  SingleHandlerConflictType conflictType;

  SingleHandlerStrategy(this.conflictType);

  @override
  onCallNoSticky(String name, A data) {
    var handler = _handlers[name];
    if (handler == null) {
      throw HandlerNotFoundException();
    }
    return handler.onCall(data);
  }

  @override
  void addHandler(Handler<A> handler) {
    super.addHandler(handler);
    if (_handlers[handler.name] == null) {
      _handlers[handler.name] = handler;
    } else {
      conflictType.addHandler(_handlers, handler);
    }
  }

  @override
  void removeHandler(Handler<A> handler) {
    _handlers[handler.name] = handler;
  }
}

enum SingleHandlerConflictType {
  /// 替换
  replace,

  /// 忽略
  ignore,

  /// 异常
  exception;

  void addHandler<A>(Map<String, Handler<A>> handlers, Handler<A> handler) {
    switch (this) {
      case SingleHandlerConflictType.replace:
        handlers[handler.name] = handler;
        break;
      case SingleHandlerConflictType.ignore:
        break;
      case SingleHandlerConflictType.exception:
        throw Exception("同名字的handler已存在");
    }
  }
}
